asterius/rts/rts.scheduler.mjs

```diff
@@ -0,0 +1,475 @@
+            this.submitCmdYield();
+            break;
+          }
+          case 2: { // ThreadInterpret
+            throw "Scheduler: unsupported ThreadInterpret";
+            this.completeTSOs.add(tid);
+            this.submitCmdYield();
+            break;
+          }
+          case 3: { // ThreadKilled
+            tso_info.ret   = 0;
+            tso_info.rstat = 2; // Killed (SchedulerStatus)
+            tso_info.promise_resolve(tid); // rts_eval* functions assume a TID is returned
+            this.completeTSOs.add(tid);
+            this.submitCmdYield();
+            break;
+          }
+          case 4: { // ThreadComplete
+            const tso_info = this.tsos.get(tid);
+            const stackobj = Number(this.memory.i64Load(tso + rtsConstants.offset_StgTSO_stackobj));
+            const sp       = Number(this.memory.i64Load(stackobj + rtsConstants.offset_StgStack_sp));
+            tso_info.ret   = Number(this.memory.i64Load(sp + 8));
+            tso_info.rstat = 1; // Success (SchedulerStatus)
+            tso_info.promise_resolve(tid); // rts_eval* functions assume a TID is returned
+            this.completeTSOs.add(tid);
+            this.submitCmdYield();
+            break;
+          }
+        }
+        break;
+      }
+      default: {
+        throw `returnFFIPromise: unsupported thread stopping reason ${reason}`;
+        break;
+      }
+    }
+  }
+
+  /**
+   * Start the scheduler
+   */
+  run(exports) {
+    exports.context.reentrancyGuard.enter(0);
+    return this.scheduler_loop(exports).then
+      ( () => { exports.context.reentrancyGuard.exit(0); }
+      , e  => {
+          // signal all the TSOs that they won't complete
+          for (var [tid,tso_info] of this.tsos) {
+            tso_info.promise_reject(`Scheduler died with: ${e.stack}`);
+          }
+          exports.context.reentrancyGuard.exit(0);
+          throw e;
+      }
+      );
+  }
+
+  /**
+   * Scheduler loop
+   */
+  async scheduler_loop(e) {
+    while (true) {
+      // read a command from the channel
+      this.running = false;
+      const cmd = await this.channel.take();
+      this.running = true;
+
+      switch (cmd.type) {
+
+        case Cmd.createThread: {
+          // call any "createThread" variant. This calls newTSO to get a fresh
+          // ThreadId.
+          const tso = e[cmd.createThread](cmd.closure);
+          const tid = this.getTSOid(tso);
+          const tso_info = this.tsos.get(tid);
+          //console.log(`Thread ${tid}: created`);
+
+          // Link the Promise returned synchronously on command submission with
+          // the TSO promise.
+          tso_info.promise.then
+            (r => cmd.resolve(r)
+            ,e => cmd.reject(e)
+            );
+
+          // Add the thread into the run-queue
+          this.enqueueTSO(tso);
+
+          break;
+        }
+
+        case Cmd.yield: {
+          if (this.runQueue.length > 0) {
+            const tid = this.runQueue.shift();
```

Comment:
For now, it's fine for us to use an array for the run queue and `.shift()` for fetching the queue head. However, `.shift()` is usually O(n) (unlike `.pop()`) and can be a problem with a large number of runnable threads in the queue, so we also need to add a ticket about a more efficient queue implementation.